### ISSUE TITLE
ANR with ParseUser.getCurrentUser()

### ISSUE THREAD

**Issue by michaelgarnerdev** [ParsePlatform/Parse-SDK-Android#390]
DISCLAIMER: I am unsure if this is actually an issue with the Parse SDK.

I have experienced this issue with 1.12.0 and 1.13.0 of the Parse SDK using bolts 1.3.0 and 1.4.0 respectively. Some of my users are experiencing an issue where our app displays a black screen and eventually ANRs and quits upon launch. I was able to get the anr traces.txt and found this:

```
"main" prio=5 tid=1 WAIT
  | group="main" sCount=1 dsCount=0 obj=0x41c1dea0 self=0x41c09768
  | sysTid=31669 nice=-11 sched=0/0 cgrp=apps handle=1074811220
  | state=S schedstat=( 0 0 0 ) utm=130 stm=12 core=1
  at java.lang.Object.wait(Native Method)
  - waiting on <0x427b26a8> (a java.lang.Object)
  at java.lang.Object.wait(Object.java:364)
  at bolts.Task.waitForCompletion(Task.java:132)
  at com.parse.ParseTaskUtils.wait(ParseTaskUtils.java:29)
  at com.parse.ParseUser.getCurrentUser(ParseUser.java:906)
  at com.parse.ParseUser.getCurrentUser(ParseUser.java:893)
  at com.ourapp.activities.LauncherActivity.onCreate(LauncherActivity.java:38)
  at android.app.Activity.performCreate(Activity.java:5458)
  at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1093)
  at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2376)
  at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2470)
  at android.app.ActivityThread.access$900(ActivityThread.java:174)
  at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1307)
  at android.os.Handler.dispatchMessage(Handler.java:102)
  at android.os.Looper.loop(Looper.java:146)
  at android.app.ActivityThread.main(ActivityThread.java:5593)
  at java.lang.reflect.Method.invokeNative(Native Method)
  at java.lang.reflect.Method.invoke(Method.java:515)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1283)
  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1099)
  at dalvik.system.NativeStart.main(Native Method)
```

This occurs during a call to:

`ParseUser user = ParseUser.getCurrentUser();` in `onCreate()` for `LauncherActivity` (our Main Launcher Activity for the app)

Parse is initialized in the Application class `onCreate()`

This is occurring only with previously authenticated users and after the app has been force closed and relaunched (also happens with an update for the same reason). If they have a fresh install (or cleared data) it does not occur. As far as OS versions and devices it varies quite a bit, but I am able to reproduce it every time with a Samsung Galaxy S4 Mini (GT-I9190) running 4.4.2 KitKat and also with a Samsung Galaxy S4 (GT-I545) running 4.4.2 KitKat. I have seen users report this problem from various manufactures and both pre-lollipop and post lollipop though. 

Any ideas??

You may read this information on SO as well:
http://stackoverflow.com/questions/35188481/parse-android-sdk-parseuser-getcurrentuser-anr

**Comment by grantland** [comment]
This is very possible, since we have to read the current user off disk if it's not cached, which would be on cold starts. We try to eliminate it by asynchronously loading it from `Parse.initialize(..)`, but your code may be trying to access it sooner.

In general, this disk I/O shouldn't be enough to cause an ANR unless the device was resource constrained, but the devices you mentioned are the worst of the crop so it could be the case here.

There's not much that can be done from the SDK since `ParseUser.getCurrentUser()` is a blocking API call, but if it's a big enough issue you could move this call to a background thread like so:

```java
Task<ParseUser> getCurrentUserAsync() {
  return Task.call(new Callable<ParseUser>() {
    public ParseUser call() {
      return ParseUser.getCurrentUser();
    }
  }, Task.BACKGROUND_THREAD);
}
```

**Comment by michaelgarnerdev** [comment]
Found the issue. We had a custom User class extending ParseUser. I had created a method called "isCurrentUser()" which is also the name of a package method in ParseUser. Lint didn't see a problem with it accordingly, but when Parse goes to retrieve the current user it calls isCurrentUser() in the process and the custom method I created was returning unexpected values to the SDK which then caused an eternal lock on the object in bolts. CRAZY.
